internal/gitutil/git_test.go

```diff
@@ -0,0 +1,213 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitutil
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"sigs.k8s.io/krew/internal/testutil"
+)
+
+func assertRepoIsCloned(t *testing.T, repoPath string) {
+	t.Helper()
+
+	ok, err := IsGitCloned(repoPath)
+
+	if err != nil {
+		t.Errorf("expected to finish correctly: %s", err)
+	}
+
+	if ok != true {
+		t.Errorf("expected to return true after cloning the repo")
+	}
+}
+
+func TestEnsureClonedClonesARepository(t *testing.T) {
+	tempDir, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	httpClonePath := tempDir.Path("krew-from-https")
+	localClonePath := tempDir.Path("krew-from-local-path")
+
+	if err := EnsureCloned("https://github.com/kubernetes-sigs/krew.git", httpClonePath); err != nil {
+		t.Errorf("http clone expected to finish correctly: %s", err)
+	}
+
+	if err := EnsureCloned(httpClonePath, localClonePath); err != nil {
+		t.Errorf("folder clone expected to finish correctly: %s", err)
+	}
+
+	assertRepoIsCloned(t, httpClonePath)
+	assertRepoIsCloned(t, localClonePath)
+}
+
+func TestEnsureClonedDoesntCloneAnExistingRepo(t *testing.T) {
+	tempDir, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	if err := EnsureCloned("https://github.com/kubernetes-sigs/krew.git", tempDir.Root()); err != nil {
+		t.Errorf("clone expected to finish correctly: %s", err)
+	}
+
+	tempDir.Write("file", []byte("create a file to ensure is not overwriten in the next clone"))
+
+	if err := EnsureCloned("https://github.com/kubernetes-sigs/krew.git", tempDir.Root()); err != nil {
+		t.Errorf("second clone expected to finish correctly: %s", err)
+	}
+
+	_, err := os.Stat(tempDir.Path("file"))
+
+	if err != nil {
+		t.Error("expected to dont perform any operation in the second clone")
+	}
+}
+
+func TestIsGitClonedWhenPathDoesntExists(t *testing.T) {
+	tempDir, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	ok, err := IsGitCloned(tempDir.Path("does-not-exist"))
+
+	if err != nil {
+		t.Errorf("expected to finish correctly: %s", err)
+	}
+
+	if ok != false {
+		t.Errorf("expected to return false on a non existing folder")
+	}
+}
+
+func TestIsGitClonedWhenIsFalse(t *testing.T) {
+	tempDir, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	ok, err := IsGitCloned(tempDir.Root())
+
+	if err != nil {
+		t.Errorf("expected to finish correctly: %s", err)
+	}
+
+	if ok != false {
+		t.Errorf("expected to return false on a folder that's not a git repo")
+	}
+}
+
+func TestIsGitClonedWhenIsTrue(t *testing.T) {
+	tempDir, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	if err := os.MkdirAll(tempDir.Path(".git"), os.ModePerm); err != nil {
+		t.Fatalf("cannot create directory %q: %s", filepath.Dir(tempDir.Path(".git")), err)
+	}
+
+	ok, err := IsGitCloned(tempDir.Root())
+
+	if err != nil {
+		t.Errorf("expected to finish correctly: %s", err)
+	}
+
+	if ok != true {
+		t.Errorf("expected to return true on a git repo")
+	}
+}
```

Comment:
you can actually merge all these into 1 test case very easily. and merge `if-if` to `if-elseif`

Reply:
the two first tests, sure I'll merge.

The last one has a custom `mkdirall` as part of the setup so I think merging them will make the code less clear? how would you merge this one with the others normally?